docs/concepts/caching.rst

```diff
@@ -0,0 +1,489 @@
+- Checkpoint and resume your dataflow in case of failures
+- Tune your dataflow (i.e., similar to hyperparameter tuning in ML) by iterating over inputs and configs
+
+This page first covers how to use caching in Hamilton then details the implementation of the caching mechanisms.
+
+.. warning::
+
+    The caching feature is under active development. If you have questions or encounter any issue, please reach out via GitHub Issues or the Hamilton Slack community server.
+
+
+Basics
+-------
+
+Simply add the ``.with_cache()`` clause to the ``Builder()`` to get started. By default, it will create in the current directory a ``hamilton_cache/`` subdirectory. ``Driver.execute()`` will execute the dataflow as usual, but it will also store metadata and results under ``hamilton_cache/`` for future use. Subsequent calls to ``Driver.execute()`` will leverage the metadata to determine if results can be loaded from disk and skip node execution when possible!
+
+
+.. code-block:: python
+
+    from hamilton import driver
+    import my_dataflow
+
+    dr = (
+        driver.Builder()
+        .with_module(my_dataflow)
+        .with_cache()
+        .build()
+    )
+
+    dr.execute([...])
+
+
+Cache to a file
+-----------------
+
+By default, caching only uses the Python standard library and relies on the ``pickle`` format to store results, `which comes with caveats <https://grantjenks.com/docs/diskcache/tutorial.html#caveats>`_ and can be impractical. Furthermore, some Python objects can't even be pickled (e.g., text tokenizers with Rust bindings).
+
+Using the Hamilton ``@tag`` decorator, you can cache a node's result to a specific file format (``JSON``, ``CSV``, ``Parquet``, etc.) and circumvent the limitations of ``pickle`` files.
+
+
+.. code-block:: python
+
+    # my_dataflow.py
+    import pandas as pd
+    from hamilton.function_modifiers import tag
+
+    def raw_data(path: str) -> pd.DataFrame:
+        return pd.read_csv(path)
+
+    @tag(cache="parquet")
+    def clean_dataset(raw_data: pd.DataFrame) -> pd.DataFrame:
+        raw_data = raw_data.fillna(0)
+        return raw_data
+
+    @tag(cache="json")
+    def statistics(clean_dataset: pd.DataFrame) -> dict:
+        return ...
+
+
+.. note::
+
+    This aims to match and replace the features of ``hamilton.experimental.h_cache.CachingGraphAdater`` but the API is preliminary.
+
+
+Resume from checkpoint
+-----------------------
+
+The caching feature stores results after each node execution. If execution fails half-way, the ``Driver`` can use the cache metadata and results to resume the run. You can think of it as automatically passing ``.execute(overrides={...})`` from disk.
+
+To resume, you can specify ``.with_cache(resume_from=...)`` which accepts a specific ``run_id`` or the string ``"latest"`` which will automatically get the results from the latest execution, whether it succeeded or failed. The ``run_id`` is automatically generated internally and can be viewed via the cache logger.
```

Comment:
This ignores fingerprinting right? 
E.g.
1. run dataflow with `latest=True`. 
2. there is nothing in the cache -- so compute everything.
3. it fails due to a code error
4. we fix the code.
5. we restart with `latest=True`.
6. since there is data in the cache, for any node names that are the same, we will load the prior value, and continue computation from where it left off.

Reply:
It doesn't ignore fingerprinting and I believe doesn't work as you expect. 

The `resume_from=` options means "resume from this specific `run_id`". If you have previous runs ("1", "2", "3", "4"), calling `resume_from="latest"` will pick the latest available `run_id` at Driver build time (i.e., `"4"`).

Node names are never used for caching. If code versions changed, this will be a cache miss. However, if your DAG has nodes
`foo -> bar -> baz`. 
1. You execute it
2. `baz` fails
3. you tweak baz
4. you rebuild the Driver with `resume_from="latest"`
5. execute again
6. execution will start at `bar` assuming its code version didn't change